### PR TITLE
Small bug fixes

### PR DESCRIPTION
**Pull request by apaszke**
* Duplicate backward roots are now handled properly
* Added support for backprop through cuDNN RNN in evaluation mode
* cuDNN dropout descriptors will be updated when RNN module's `.dropout` attribute is changed. Additionally, they are now lazily initialized, so the descriptors won't even be seeded if `dropout == 0` (cuDNN is fine with that, so I assume it's ok).

@ngimel can you please take a look at the changes in the backend?

### CODE REVIEW

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -73,14 +74,33 @@ auto Engine::backward(const variable_list& variables,
...
-      if (var->creator->requires_grad) {
-        GradBuffer buf(var->creator->num_outputs);
-        buf.addGrad(var->output_nr, Variable::of(std::move(grad)));
-        ready.emplace_front(var->creator, std::move(buf));
+      auto& creator = var->creator;
+      auto it = creator_grad.find(creator);
+      if (it == creator_grad.end()) {
+        GradBuffer* buf = nullptr;
+        if (creator->requires_grad) {
+          buf = new GradBuffer(creator->num_outputs);
```
Comment (colesbury) [inline torch/csrc/autograd/engine.cpp]: This looks like it leaks memory. I don't see where buf is freed.

File: torch/csrc/autograd/engine.cpp [hunk 2]
```diff
@@ -59,6 +59,7 @@ auto Engine::backward(const variable_list& variables,
                       tensor_list& grad_variables,
                       bool retain_variables) -> void {
   function_queue creators;
+  std::unordered_map<std::shared_ptr<Function>, GradBuffer*> creator_grad;
```
Comment (colesbury) [inline torch/csrc/autograd/engine.cpp]: What problem is this solving?

Comment (apaszke) [inline torch/csrc/autograd/engine.cpp]: Many variables having the same creator.

File: torch/csrc/autograd/engine.cpp (hunk 1 above)
Comment (apaszke) [inline torch/csrc/autograd/engine.cpp]: Yeah, you're right...

File: torch/csrc/autograd/engine.cpp (hunk 2 above)
Comment (apaszke) [inline torch/csrc/autograd/engine.cpp]: Right now they'll be added to the `ready` queue multiple times, with different grads. After this change they're first gathered in that map, with duplicate entries having their grad accumulated, and only then are they added to the queue. You can see that I removed a `+ 0` hack in nn tests that was necessary because of the bug.

File: torch/backends/cudnn/__init__.py [hunk 3]
```diff
@@ -241,24 +241,42 @@ class DropoutDescriptor(object):
...
+            self._set(dropout, seed)
+
+    def _set(self, dropout, seed):
+        if self.state is None and dropout > 0:
+            dropout_states_size = ctypes.c_long()
+            check_error(lib.cudnnDropoutGetStatesSize(
+                self.handle,
+                ctypes.byref(dropout_states_size)))
+            self.state = torch.cuda.ByteTensor(dropout_states_size.value)
+            state_ptr = self.state.data_ptr()
```
Comment (colesbury) [inline torch/backends/cudnn/__init__.py]: Is state_ptr a local variable?

File: torch/backends/cudnn/__init__.py [hunk 4]
```diff
@@ -241,24 +241,42 @@ class DropoutDescriptor(object):
...
+            state_size = 0
 
         check_error(lib.cudnnSetDropoutDescriptor(
             self,
-            handle,
+            self.handle,
             ctypes.c_float(dropout),
-            ctypes.c_void_p(self.state.data_ptr()),
-            ctypes.c_size_t(self.state.size(0)),
+            ctypes.c_void_p(state_ptr),
```
Comment (colesbury) [inline torch/backends/cudnn/__init__.py]: Why not just `self.state.data_ptr()` and `self.state.size(0)`? (fewer things that need to be kept consistent)

File: torch/csrc/autograd/engine.cpp [hunk 5]
```diff
@@ -59,6 +59,7 @@ auto Engine::backward(const variable_list& variables,
                       tensor_list& grad_variables,
                       bool retain_variables) -> void {
   function_queue creators;
+  std::unordered_map<std::shared_ptr<Function>, std::unique_ptr<GradBuffer>> creator_grad;
```
Comment (colesbury) [inline torch/csrc/autograd/engine.cpp]: Should the key just be `Function*`? This is probably fine too.

File: torch/csrc/autograd/engine.cpp [hunk 6]
```diff
@@ -73,14 +74,33 @@ auto Engine::backward(const variable_list& variables,
         did_leaf_backward = true;
       }
     } else {
-      creators.push_back(var->creator.get());
-      if (var->creator->requires_grad) {
-        GradBuffer buf(var->creator->num_outputs);
-        buf.addGrad(var->output_nr, Variable::of(std::move(grad)));
-        ready.emplace_front(var->creator, std::move(buf));
+      auto& creator = var->creator;
+      auto it = creator_grad.find(creator);
```
Comment (colesbury) [inline torch/csrc/autograd/engine.cpp]: You can probably combine the two cases a bit since unique_ptr is default constructible by doing something like:

```
if (var->requires_grad) {
 auto& buf = creator_grad[creator]
 if (!buf) {
   buf.reset(new ...)
 }
 buf->addGrad(...)
}
```

File: torch/csrc/autograd/engine.cpp [hunk 7]
```diff
@@ -73,14 +74,33 @@ auto Engine::backward(const variable_list& variables,
...
+        if (creator->requires_grad) {
+          auto& buf = it->second;
+          buf->addGrad(var->output_nr, Variable::of(std::move(grad)));
+        }
       }
     }
   }
+  for (auto& entry: creator_grad) {
+    const auto& creator = entry.first;
+    auto& buf = entry.second; // WARNING: this is nullptr if !creator->requires_grad
```
Comment (colesbury) [inline torch/csrc/autograd/engine.cpp]: It looks like creator_grad only has entries for creators which require grad. It doesn't seem like this warning is accurate.

**Review by colesbury — COMMENTED** [review]
The variable handling in engine feels like it's getting more complicated than it ought to be, since we're doing similar things with GradBuffer for variables and functions. There's probably a refactor that would simplify these things but perhaps that's best left for a later commit.

File: torch/backends/cudnn/__init__.py (hunk 3 above)
Comment (apaszke) [inline torch/backends/cudnn/__init__.py]: Yes?

File: torch/backends/cudnn/__init__.py (hunk 4 above)
Comment (apaszke) [inline torch/backends/cudnn/__init__.py]: because sometimes `self.state` can be None. If you create a dropout descriptor with `dropout == 0`, it won't even be initialized.

File: torch/csrc/autograd/engine.cpp (hunk 5 above)
Comment (apaszke) [inline torch/csrc/autograd/engine.cpp]: Can you unpack a shared pointer and re-pack it again later?

File: torch/csrc/autograd/engine.cpp (hunk 7 above)
Comment (apaszke) [inline torch/csrc/autograd/engine.cpp]: It has entries for those which don't require grad, but the buffer is `nullptr`. [This line](https://github.com/pytorch/pytorch/pull/886/files#diff-01f226b610655c49b1e52eddf9ea4126R85) is not guarded by a `requires_grad` check.

File: torch/backends/cudnn/__init__.py (hunk 3 above)
Comment (colesbury) [inline torch/backends/cudnn/__init__.py]: it looks like it's assigned but never used

Comment (colesbury) [inline torch/backends/cudnn/__init__.py]: Oh these are the same function, nm

File: torch/csrc/autograd/engine.cpp (hunk 5 above)
Comment (apaszke) [inline torch/csrc/autograd/engine.cpp]: No you can't. I later need to insert shared pointers to `ready`, so I need to keep them in the map too.

File: torch/backends/cudnn/__init__.py (hunk 4 above)
Comment (colesbury) [inline torch/backends/cudnn/__init__.py]: Never mind i thought this call was in a separate function

**Review by colesbury — APPROVED** [review]
